### PR TITLE
Concurrency 3

### PR DESCRIPTION
**Pull request by jsinglet**
## Description

Implements Concurrency3.

## Change request type

 - [ ] Release or process automation (GitHub workflows, internal scripts)
 - [ ] Internal documentation
 - [ ] External documentation
 - [x] Query files (`.ql`, `.qll`, `.qls` or unit tests)
 - [ ] External scripts (analysis report or other code shipped as part of a release)

## Rules with added or modified queries

 - [ ] No rules added
 - [x] Queries have been added for the following rules:
     - CON31-C
     - CON41-C
     - CON38-C
 - [ ] Queries have been modified for the following rules:
     - CON50-C

## Release change checklist

A change note ([development_handbook.md#change-notes](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#change-notes)) is required for any pull request which modifies:

 - The structure or layout of the release artifacts.
 - The evaluation performance (memory, execution time) of an existing query.
 - The results of an existing query in any circumstance.

If you are only adding new rule queries, a change note is not required.

_**Author:**_ Is a change note required? 
  - [x] Yes
  - [ ] No

_**Reviewer:**_ Confirm that either a change note is not required or the change note is required and has been added.
  - [x] Confirmed

## Query development review checklist

For PRs that add new queries or modify existing queries, the following checklist should be completed by both the author and reviewer:

### Author

 - [ ] Have all the relevant rule package description files been checked in?
 - [ ] Have you verified that the metadata properties of each new query is set appropriately?
 - [ ] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [ ] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [ ] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [ ] Does the query have an appropriate level of in-query comments/documentation?
 - [ ] Have you considered/identified possible edge cases?
 - [ ] Does the query not reinvent features in the standard library?
 - [ ] Can the query be simplified further (not golfed!)

### Reviewer 

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### CODE REVIEW

**Review by kraiouchkine — CHANGES_REQUESTED** [review]
Looks great other than the items discussed:
1) The shared query comment is the default one
2) cpp/common/test/rules/donotallowamutextogooutofscopewhilelocked/test.cpp and the associated query currently raises false-positives in cases where thread/synchronization functions are in wrappers. In my opinion, adding a test case to demonstrate this case and documenting it as a possible future enhancement suffices 👍

File: c/cert/test/rules/CON41-C/test.c [hunk 1]
```diff
@@ -0,0 +1,49 @@
+#include "stdatomic.h"
```
[the remark this answers is not in the document]
Comment (jsinglet) [reply c/cert/test/rules/CON41-C/test.c:1]: Thanks for all of this @kraiouchkine - I've added this feedback to https://github.com/github/codeql-coding-standards/issues/48

File: cpp/common/src/codingstandards/cpp/rules/preservesafetywhenusingconditionvariables/PreserveSafetyWhenUsingConditionVariables.qll [hunk 2]
```diff
@@ -0,0 +1,73 @@
+/**
+ * Provides a library which includes a `problems` predicate for reporting....
```
[the remark this answers is not in the document]
Comment (jsinglet) [reply cpp/common/src/codingstandards/cpp/rules/preservesafetywhenusingconditionvariables/PreserveSafetyWhenUsingConditionVariables.qll:2]: Fixed!

**Comment by jsinglet** [comment]
Test case added to cpp/common/test/rules/donotallowamutextogooutofscopewhilelocked/test.cpp @kraiouchkine

**Comment by jsinglet** [comment]
@kraiouchkine  ready for another 👁️

File: cpp/common/test/rules/donotallowamutextogooutofscopewhilelocked/test.cpp [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+void join_threads(std::thread *threads) {
+  for (int i = 0; i < 5; ++i) {
+    threads[i].join();
+  }
+}
+
+void f11() {
+  std::thread threads[5];
+  std::mutex m1; // NON_COMPLIANT (FALSE POSITIVE)
```
Comment (kraiouchkine) [inline cpp/common/test/rules/donotallowamutextogooutofscopewhilelocked/test.cpp:134]:
```suggestion
  std::mutex m1; // COMPLIANT[FALSE_POSITIVE]
```
The comment doesn't conform with the rest of our test-cases.

**Review by kraiouchkine — CHANGES_REQUESTED** [review]

**Review by kraiouchkine — APPROVED** [review]
Fixed the NON_COMPLIANT comment to COMPLIANT[FALSE_POSITIVE]. 
Otherwise LGTM, thanks for the changes!

**Pull request merged**